### ISSUE TITLE
Remapping option for channel 5 in SBUS

### ISSUE THREAD

**Issue by Kjblootens** [ExpressLRS/ExpressLRS#2584]
### Discussed in https://github.com/ExpressLRS/ExpressLRS/discussions/2479

<div type='discussions-op-text'>

<sup>Originally posted by **apiening** November 19, 2023</sup>
I have an airplane from **FMS** which has a **stabilizer** system (**Reflex V2**) which accepts **SBUS**. I enabled SBUS output for my RX (**ELRS 3.3**) and the control link works fine.
However, this stabilizer uses channel 5 with a three position mode switch to select between stabilizing modes (auto level, optimized, off). So when I configure it this way, I also disarm when I choose auto level stabilized mode.

I can assign the **PWM** outputs freely with **ELRS**, but is there a chance to **remap channel 5** for the **SBUS** output?</div>

**Comment by stevewitwitty** [comment]
I use SBUS with a Helicopter with the same issue. Helicopter uses channel 5 for stabilization mode. I got around this issue by going into the Express LRS lua script and changing the packet rate to 333Hz and then changing the switch mode to 16ch Rate/2. This then allows full resolution on channel 5 and the helicopter's stabilization mode switches on and off via a switch as needed. You can read further info here: https://www.expresslrs.org/software/switch-config/?h=switch

**Comment by stevewitwitty** [comment]
Forgot to say, there is a video which i found that helped me with this:

https://www.youtube.com/live/ARCfafma1rM?si=6MxiHAZJrRpC9g_N

Credit to RC Video Reviews for the video

**Comment by CapnBry** [comment]
A better solution is to allow remapping of all channels on the receiver. Not just SBUS not just CH5. This would let the user decide not only where CH5 ends up coming out, but also how the other channels are shifted.

**Comment by Kjblootens** [comment]
CapnBry. Yes that's implied here but still a good call out.

The problem only really exists for CH5 though. For any other issues there are very viable workarounds in terms of channel remapping on the TX.

The only viable workaround here are 16/2 Full modes where CH5 is not a prioritized channel along with CH1-4 as ALL packets get deprioritized (as per the switch config page, 16/2 modes have no lines bolded.) https://www.expresslrs.org/software/switch-config/

The ask is to enable serial output remapping (any serial protocol) so we can avail of the benefits of other switch modes while maintaining use of CH5 for e.g. flight mode switching where the target system (stabilizer) cannot be modified.

**Comment by CapnBry** [comment]
Yeah that's what I was saying, remap all channels to any order in the webui like you can do for pwm outputs.

I don't think we'll ever "deprioritize CH5" because it is how thousands of people have their stuff set up currently, but remapping the channels means that it doesn't matter because you can have the output CH5 be whatever, and 16ch mode (which kinda sucks) won't be as used. I think the major reason people use 16 channels at half rate is just for CH5.

**Comment by Gerold68** [comment]
CapnBry,
>I think the major reason people use 16 channels at half rate is just for CH5.

You underestimate the gilder guys that use more than 8 Channels! 16/2 is made for us ;-)

**Comment by CapnBry** [comment]
> You underestimate the gilder guys that use more than 8 Channels! 16/2 is made for us ;-)

I hear ya! But CH5 is still going to be ARM no matter what, until it gets removed (in 2030 maybe?). Until then you'll still only get 15 channels unless you're not using of the things that rely on CH5, but you can do that currently without any ExpressLRS changes. Unfortunately, the proposed RX-side mapping won't work with any of the glider scripts that rely on the handset's channel order. :/

**Comment by ParkJeff9** [comment]
My main use case is fixed wing. and I would like to add my support for channel re-mapping in SBUS and other serial outputs. I don't use the workaround because, I believe, not using ch 5 for arming messes up dynamic power in the Tx. Better if the rx can map another channel to the ch 5 position in the serial data stream.

**Comment by mha1** [comment]
> My main use case is fixed wing. and I would like to add my support for channel re-mapping in SBUS and other serial outputs. I don't use the workaround because, I believe, not using ch 5 for arming messes up dynamic power in the Tx. Better if the rx can map another channel to the ch 5 position in the serial data stream.

V4 and EdgeTX 2.11 will solve the CH5 SBUS inconvenience. Read through https://github.com/ExpressLRS/ExpressLRS/pull/3008#issue-2613672057 for more info on the new arming mode. This will get you channel 5 as full proportional channel in all full-res modes and also for SBUS output while still being able to independently arm the module. No need for remapping or fiddling with ELRS code to remap SBUS.

To play with this use Configurator to build and flash branch master for your rx and tx module. And make sure to flash a recent EdgeTX nightly to your EdgeTX radio (e.g. https://github.com/EdgeTX/edgetx/releases/download/nightly/edgetx-firmware-nightly-0ea41f21.zip)

**Comment by ParkJeff9** [comment]
> V4 and EdgeTX 2.11 will solve the CH5 SBUS inconvenience. Read through [#3008 (comment)](https://github.com/ExpressLRS/ExpressLRS/pull/3008#issue-2613672057) for more info on the new arming mode. This will get you channel 5 as full proportional channel in all full-res modes and also for SBUS output while still being able to independently arm the module. No need for remapping or fiddling with ELRS code to remap SBUS.

That would be excellent. Thank you.

**Comment by airbass70** [comment]
> V4 and EdgeTX 2.11 will solve the CH5 SBUS inconvenience. Read through [#3008 (comment)](https://github.com/ExpressLRS/ExpressLRS/pull/3008#issue-2613672057) for more info on the new arming mode. This will get you channel 5 as full proportional channel in all full-res modes and also for SBUS output while still being able to independently arm the module. No need for remapping or fiddling with ELRS code to remap SBUS.
> 
> To play with this use Configurator to build and flash branch master for your rx and tx module. And make sure to flash a recent EdgeTX nightly to your EdgeTX radio (e.g. https://github.com/EdgeTX/edgetx/releases/download/nightly/edgetx-firmware-nightly-0ea41f21.zip)

When a new tag gets released, does it originate to the current branch master? Otherwise, can some commits be excluded? 
example: PR 3008 was merged with master branch in nov 2024 but it seems to be not included yet in the later tags (I cant find any of its commits in the 3.6). On the other hand, PR 3008 comes with a label V4.0 which let me think it is possible to merge it and keep it on hold. Can somebody explain how does it work?  Sorry, I know this is git dumb question. The reason being I am learning github and waiting for CH5 full res on an official release. Thank you developers for understanding.

**Comment by CapnBry** [comment]
> When a new tag gets released, does it originate to the current branch master? Otherwise, can some commits be excluded? example: PR 3008 was merged with master branch in nov 2024 but it seems to be not included yet in the later tags (I cant find any of its commits in the 3.6). On the other hand, PR 3008 comes with a label V4.0 which let me think it is possible to merge it and keep it on hold. Can somebody explain how does it work? Sorry, I know this is git dumb question. The reason being I am learning github and waiting for CH5 full res on an official release. Thank you developers for understanding.

Not a dumb question at all, because it depends on the project how pulls are integrated into releases! 

We have two paths of active development:
* master branch - This is there #3008 was merged to and will become the 4.0 release. While we were still working on 3.x stuff only, master branch also was the source of 3.x releases. Once the switch was made to OTA 4, master became incompatible with 3.x releases. All new features go to master.
* 3.x.x-maintenance branch - This is where bug fixes go generally. The idea is that we can keep working on master and if there's a bug found in e.g. 3.4.0, 3.x.x-maintenance is where we'd put the fix, so we can put out a 3.4.1 without having to worry about stuff in master that would be for 3.5.0 possibly breaking 3.4.x users. Every so often, changes from 3.x.x-maint are merged back to master so it also has those bug fixes for 3.5.0.

There's no 3.7.0 coming, because master is 4.0.0, so there is still some activity on 3.x.x-maintenance for a couple bug fixes for 3.6.0 but there has not been a release that includes #3008 because it was a feature and there hasn't been a feature release since it was merged. This feature also relies on EdgeTX support, and I am not sure if that's been released either?

Hope that clears it up!

**Comment by mha1** [comment]
https://github.com/ExpressLRS/ExpressLRS/pull/3008 is supported by EdgeTX 2.11

**Comment by airbass70** [comment]
> Hope that clears it up!

yes! it does. Thank you very much sir.

**Comment by J-Sorenson** [comment]
Isn't this essentially resolved with https://github.com/ExpressLRS/ExpressLRS/pull/3008 for V4.0?  Seems like it can be closed.

**Comment by r1000ru** [comment]
Is this decision reflected in the CRSF protocol? Or is it invented by the developers of ELRS and EdgeTX and will not be implemented by hardware manufacturers on other operating systems that support the official CRSF?

**Comment by CapnBry** [comment]
It is only implemented in EdgeTX EDIT: I am incorrect, see mha1's reply. Any other handset operating system which wants that feature must implement it as well. Regardless of its official CRSF status, there's no way to conjure up an additional flag or channel without a change to the handset software.

**Comment by mha1** [comment]
@r1000ru Ethos also decided to implement what's required for #3008. What other OS are you thinking of? Anyhow, not implementing the handset side will maintain the status quo AUX1/CH5 arming.

**Comment by Patronics** [comment]
This doesn't seem to have been resolved with ELRS 4.0.0. I'm testing an SBUS output which seems to still be tied to the arm/disarm signal, (even after making the settings change in edgetx) as confirmed with a logic analyzer.
Here's the TX configuration
![Image](https://github.com/user-attachments/assets/e779d748-5fd1-4b7e-ac8c-f777d8cb02b0)
And here's a recording in saelea logic demonstrating that channel 5 is still being toggled when I press the arm/disarm switch.
[SBUS arm Switch toggle recording.sal.zip](https://github.com/user-attachments/files/26109607/SBUS.arm.Switch.toggle.recording.sal.zip)

In the edgetx outputs view, there's no indication of any changes to channel 5 when I arm or disarm, and switch configured input for channel 5 does show as changing the channel 5 output, though that seems to not really be the case.

Tested with Radiomaster TX15 Running Edgetx 2.12.0-RC4.

**Comment by Kjblootens** [comment]
> This doesn't seem to have been resolved with ELRS 4.0.0. I'm testing an SBUS output which seems to still be tied to the arm/disarm signal, (even after making the settings change in edgetx) as confirmed with a logic analyzer. Here's the TX configuration ![Image](https://github.com/user-attachments/assets/e779d748-5fd1-4b7e-ac8c-f777d8cb02b0) And here's a recording in saelea logic demonstrating that channel 5 is still being toggled when I press the arm/disarm switch. [SBUS arm Switch toggle recording.sal.zip](https://github.com/user-attachments/files/26109607/SBUS.arm.Switch.toggle.recording.sal.zip)
> 
> In the edgetx outputs view, there's no indication of any changes to channel 5 when I arm or disarm, and switch configured input for channel 5 does show as changing the channel 5 output, though that seems to not really be the case.
> 
> Tested with Radiomaster TX15 Running Edgetx 2.12.0-RC4.

Hi for fear of sounding silly. I don't think channel 5 "gets toggled" rather needs to be toggled.

Are you sure there isnt a ch5 arm mix in your radio?

Ch5 NEEDED to be used for arming to enable some ELRS functionality. It didn't flip CH5 based on an arm signal.

**Comment by mha1** [comment]
> This doesn't seem to have been resolved with ELRS 4.0.0. I'm testing an SBUS output which seems to still be tied to the arm/disarm signal, (even after making the settings change in edgetx) as confirmed with a logic analyzer. Here's the TX configuration ![Image](https://github.com/user-attachments/assets/e779d748-5fd1-4b7e-ac8c-f777d8cb02b0) And here's a recording in saelea logic demonstrating that channel 5 is still being toggled when I press the arm/disarm switch. [SBUS arm Switch toggle recording.sal.zip](https://github.com/user-attachments/files/26109607/SBUS.arm.Switch.toggle.recording.sal.zip)
> 
> In the edgetx outputs view, there's no indication of any changes to channel 5 when I arm or disarm, and switch configured input for channel 5 does show as changing the channel 5 output, though that seems to not really be the case.
> 
> Tested with Radiomaster TX15 Running Edgetx 2.12.0-RC4.

SBUS outputs the full proportional value of EdgeTX mixer for CH5 if:
- ELRS version is min 4.0.0
- EdgeTX version is min 2.11.0
- Arm using is set to Switch
- packet rate is set to 100Hz full res or 333Hz full res

**Comment by Patronics** [comment]
Yep, I just stumbled across the solution myself, it's just as you say @mha1, it just needed to be switched to the 333hz full mode. Thanks! It probably would be good to have a warning in the configuration screen (as in above photo) if that isn't the case, I'm sure a lot of people will stumble across that same issue! 

I'm not sure if that screen is defined within ELRS.lua or if I should open an issue with EdgeTX about it?

**Comment by OzDruiD** [comment]
> It probably would be good to have a warning in the configuration screen (as in above photo) if that isn't the case, I'm sure a lot of people will stumble across that same issue!
> 

IMHO (as just a humble user) Its hard to make a balance.   You would need to work out some logic (not too hard) but the never ending story of where do you draw the line in terms of user prompts.

Logic
1. if packet mode is not full rez, then warn that whilst they have selected switch mode - its technically still impacting RC channel 5
2. if packet mode IS full rez, then warn they need to clean up any old RC channel 5 mixes they may have - eg they updated from ELRS v3.xxx and now implemented switch mode arming, but previously had a chan5 mix for arm - which should \ now can be "Freed" up.
3.  It cannot "prevent people" ie limiting the drop down to only allow switch mode to be chosen if on full rez, for many reasons, setting up a new not active mode, people wanting BOTH, etc etc - 

Warnings
4. what about people then that have a new model setup or dont have a RC chan 5 mix yet,  #2 would confuse
5. what about drone\channel 5 users that leave it on chan 5 and dont activate swtich - for consisentency do we also add a warning NOT to hve some kind of non arm related mix on rc chan 5 - expecting like stabliasation modes or other flight modes to work correcctly and not disable other thing on the craft linked to rc chan 5.
6. Other communits like Gliders, fixed wing,  BVLOS, surface - impossible to think of all avenues and 'please' everyone, so i think the ELRS wiki \ guide on ARMing is actually very good - just needs people to be aware and READ and follow - but so would pop up warnings....

So TLDR, its a good idea - limit user missunderstanding, but how far do you take it ..

**Comment by J-Sorenson** [comment]
This is becoming a similar topic to how you need "full" mode in order to get the full servo PWM range output.  There is only so much on-boarding you can do.   Sure, if we could turn back time, there's a number of things we could've done differently that would make it cleaner, but the devs are trying not to make a mess of existing systems (especially with the huge number of multirotors).

Only now is ELRS starting to get "real" adoption amongst the fixed-wings.  We just need to make it clear that we are to use 100Hz(full) or 333Hz(full) and to use 4.0 or newer.  That's been the messaging in RCGroups.  Let's avoid over-complicating the codebase going forward.

If anything, I think it's Edge-TX that needs to handle the user guidance.  They have the GUI that defines if ARM is switch or channel.
